add "AND gate" EnsureOrigin implementation

Mr-Leshiy opened the pull request:
This an implementation of this issue https://github.com/paritytech/substrate/issues/3059.

bkchr commented on frame/system/src/lib.rs:
```diff
@@ -953,6 +953,28 @@ impl<
+
+	#[cfg(feature = "runtime-benchmarks")]
+	fn successful_origin() -> O {
+		L::successful_origin()
```
This is "wrong". Because a successful origin of `L` doesn't mean it will be successful for `R`.

thiolliere replied:
Indeed, I see no easy implementation, maybe we can start creating a origin variant only when the feature runtime-benchmarks is activated.
Like frame_system would define a new variant `AlwaysSuccess`.
But I don't very like this solution.

Mr-Leshiy replied:
Yea, have missed this case. Updating implementation.

xlc reviewed (commented):
I don't think they way this implemented is something what people wants.

You won't be able to use this for any actual useful configuration.

Take an example, we want to setup a original that requires a council member approval and a technical committee member approval. How do we use this to support such case?

thiolliere commented:
> I don't think they way this implemented is something what people wants.
> 
> You won't be able to use this for any actual useful configuration.
> 
> Take an example, we want to setup a original that requires a council member approval and a technical committee member approval. How do we use this to support such case?

Yes in case of EnsureOrigin implementation which expect a specific variant there is usecase but maybe there can be some more general implementation like someone could give some specific priviledge level of origin. like say 2/3 of council is priviledge 2 and 99% of democracy is priviledge 2 also.
Then someone could write: EnsureOrigin = Priviledge2 and from council.

KiChjang commented on frame/system/src/lib.rs:
```diff
@@ -953,6 +953,28 @@ impl<
+/// The "AND gate" implementation of `EnsureOrigin`.
+///
+/// Origin check will pass if `L` and `R` origin check passes. `L` is tested first.
+pub struct EnsureBothOf<AccountId, L, R>(sp_std::marker::PhantomData<(AccountId, L, R)>);
```
I've actually been thinking about whether it's a good thing to introduce such a new struct, or simply implement `EnsureOrigin` on tuples using impl_for_tuple. That way we're not constrained on just 2 origins, but can specify as many `EnsureOrigin`s as we want.

Mr-Leshiy replied:
I am not sure that it will be better than such implementation of `EnsureBothOf` or `EnsureOneOf`, because it will not so flexible as with the usage if these ones.

KiChjang replied:
I'm sorry, but why would it be less flexible? Compare `EnsureBothOf<AccountId, EnsureRoot, EnsureBothOf<AccountId, EnsureSigned, EnsureFounder>>` and `(EnsureRoot, EnsureSigned, EnsureFounder)`, you're telling me that the former is more flexible than the latter? It does not look so to me.

Mr-Leshiy replied:
At this case yes, but I am looking into possibility to combine `EnsureBothOf` with `EnsureOneOf` and leave an under stable interface.
But I like your idea how to make such struct with the different amount arguments for both of them `EnsureBothOf`, `EnsureOneOf`. Will try to implement it.

kianenigma replied:
Using tuples to imply `Or` is 100% more readable than nested types.

thiolliere replied:
note that we can have EnsureOneOf with as maybe generic as we want. no need for nested stuff.
We can create a type with default generic
```rust
pub struct EnsureOneOf<A, B, C=ImpossibleOrigin, D=ImpossibleOrigin, ...>
```
I feel like EnsureOneOf is more clear than a tuple personally

Mr-Leshiy commented:
@thiolliere , @xlc , I am little bit confusing. Can you pls little bit more elaborate what cases do you consider ?
According to this implementation I just following the description of the issue and made it as simple as possible.

xlc commented:
> Since the origin cannot be collected together, it would necessitate a module which retained state over dispatch/proposal hashes, allowing one origin to approve a hash first, and the other origin to approve it later causing the dispatch.

thiolliere commented:
> > Since the origin cannot be collected together, it would necessitate a module which retained state over dispatch/proposal hashes, allowing one origin to approve a hash first, and the other origin to approve it later causing the dispatch.


> I am little bit confusing. Can you pls little bit more elaborate what cases do you consider ? According to this implementation I just following the description of the issue and made it as simple as possible.
> 
> Am I right that the problem that the trait bound for the `O` type is specialized to the `RawOrigin` ? So need to make these implementation more general ?

Ok I finally understood the concern of xlc, the current implementation of the PR doesn't follow what the issue is about. The issue is not about providing a type which ensures that 2 constraint are ensured on the same origin, but instead providing an ensure origin implementation which succeed only when 2 different origin have try to dispatch a call.
So the issue is very different from the current implementation.

thiolliere commented on frame/support/src/traits/dispatch.rs:
```diff
@@ -38,6 +38,55 @@ pub trait EnsureOrigin<OuterOrigin> {
+/// The "OR gate" implementation of `EnsureOrigin`.
+///
+/// Origin check will pass if `L` or `R` origin check passes. `L` is tested first.
+pub struct EnsureOneOf<L, R>(sp_std::marker::PhantomData<(L, R)>);
```
I liked this refactor which gets rid of unnecessary generics. maybe we can isolate in another PR if this one doesn't get merged

Mr-Leshiy replied:
Yes, sure. Will move this code into the another PR.

thiolliere commented on frame/support/src/traits/dispatch.rs:
```diff
@@ -38,6 +38,55 @@ pub trait EnsureOrigin<OuterOrigin> {
+					|_| {
+						panic!(
+							"It is impossible to generate successful origin from the provided ones"
+						)
```
I understand that it is impossible to generate a successful origin in this implementation, but it is not necessary impossible to generate a successful origin which ensure both L and R ensures are successful.

Personally I think it is better to not provide EnsureBothOf if we can't implement it correctly generically.
Instead people will need to implement it manually and give a proper successful origin in the implementation.

Mr-Leshiy commented:
@xlc , @thiolliere thanks guys for explanation, it seems I have got what do you mean, will try to do the correct implementation.😊

Mr-Leshiy commented:
@thiolliere , @xlc , I have added new implementation of the AND gate with the validation from the different origins, but also remain implementation from the one origin to be respect with the previous implementation of the OR gate. Also added a new implementation of the OR gate from the different origins. 
Waiting for your review 😊

xlc commented:
I am completely lost. I don't understand what are you trying to achieve here.

I have already quoted Gav's words from the original issue, but I guess I need to emphasize it again.

> Since the origin cannot be collected together, it would necessitate a module which retained state over dispatch/proposal hashes, allowing one origin to approve a hash first, and the other origin to approve it later causing the dispatch.

i.e. You have to write a pallet to do this. It will be somewhat similar to 2/2 multisg.

thiolliere commented on frame/support/src/traits/dispatch.rs:
```diff
@@ -115,13 +116,90 @@ impl<OuterOrigin, L: EnsureOrigin<OuterOrigin>, R: EnsureOrigin<OuterOrigin>>
+pub struct EnsureOneOfDifferentOrigins<L, R>(sp_std::marker::PhantomData<(L, R)>);
+
+impl<OuterOrigin1, OuterOrigin2, L: EnsureOrigin<OuterOrigin1>, R: EnsureOrigin<OuterOrigin2>>
+	EnsureOrigin<(OuterOrigin1, OuterOrigin2)> for EnsureOneOfDifferentOrigins<L, R>
```
The implementation of `EnsureOrigin<(OneOrigin, AnotherOrigin)>` seems not useful, there is only one origin type defined for the runtime.
I don't see how there could be 2 different origin type.

thiolliere reviewed (changes requested):
Yes Xlc is correct, to implement what the issue wrote about we need another pallet.

kianenigma commented:
ping @Mr-Leshiy

Mr-Leshiy commented:
@thiolliere , @kianenigma, @xlc Sorry for the delay guys, have a lot of mess previous week. 
It seems firstly I have misunderstood you, I have followed this comment, so that is why I have done what I have done 😊

> The issue is not about providing a type which ensures that 2 constraint are ensured on the same origin, but instead providing an ensure origin implementation which succeed only when 2 different origin have tried to dispatch a call.

Will try to implement what do you mean.
Should it be a new pallet as a part of `frame` or where it is better to place it ?

xlc commented:
It will be a few frame pallet. It could also be part of the frame-utility pallet but I am not 100% sure if that's a good idea.

Mr-Leshiy commented:
Still in progress
